modulemd/v2/modulemd-defaults-v1.c

```diff
@@ -47,6 +48,95 @@ G_DEFINE_TYPE (ModulemdDefaultsV1,
                MODULEMD_TYPE_DEFAULTS)
 
 
+static gboolean
+modulemd_defaults_v1_equals (ModulemdDefaults *self_1, ModulemdDefaults *self_2, GError **error)
+{
+  ModulemdDefaultsV1 *v1_self_1, *v1_self_2;
+  v1_self_1 = NULL;
+  v1_self_2 = NULL;
+  GHashTableIter iter_1, iter_2;
+  gpointer key_1, key_2, value_1, value_2;
+  GHashTable *profile_defaults_1, *profile_defaults_2;
+  GHashTable *intent_default_profiles_1, *intent_default_profiles_2;
+
+
+  g_return_val_if_fail (MODULEMD_IS_DEFAULTS_V1 (self_1), NULL);
+  v1_self_1 = MODULEMD_DEFAULTS_V1 (self_1);  
+  g_return_val_if_fail (MODULEMD_IS_DEFAULTS_V1 (self_2), NULL);
+  v1_self_2 = MODULEMD_DEFAULTS_V1 (self_2);  
+
+  if (!MODULEMD_DEFAULTS_CLASS (modulemd_defaults_v1_parent_class)
+             ->equals (self_1, error);)
+    {
+      return FALSE;
+    }
+
+  if (!MODULEMD_DEFAULTS_CLASS (modulemd_defaults_v1_parent_class)
+             ->equals (self_2, error);)
+    {
+      return FALSE;
+    }
+
+  if(g_strcmp0 (v1_self_1->intent, v1_self_2->intent))
+    {
+      return FALSE;
+    }
+
+  const gchar *default_stream_1 =
+    modulemd_defaults_v1_get_default_stream(v1_self_1, v1_self_1->intent);
+  const gchar *default_stream_2 =
+    modulemd_defaults_v1_get_default_stream(v1_self_2, v1_self_2->intent);
+  if(g_strcmp0 (default_stream_1, default_stream_2))
+    {
+      return FALSE;
+    }  
+
+  profile_defaults_1 = v1_self_1->profile_defaults;
+  profile_defaults_2 = v1_self_2->profile_defaults;
+
+  if (g_hash_table_size (profile_defaults_1) != g_hash_table_size(profile_defaults_2))
+    {
+      return FALSE;
+    }
+  g_hash_table_iter_init (&iter_1, profile_defaults_1);
+  g_hash_table_iter_init (&iter_2, profile_defaults_2);
+
+  while (g_hash_table_iter_next (&iter_1, &key_1, &value_1) && g_hash_table_iter_next (&iter_2, &key_2, &value_2))
+    {
+      if(g_strcmp0((gchar *)key_1, (gchar *)key_1))
```

Comment:
You're comparing `key_1` to `key_1` here. However, you can't actually do what you were trying to do because there is actually no guarantee that you will get the keys from the two hash tables in the same order.

What I'd recommend doing is first getting the ordered set of keys using `modulemd_ordered_str_keys()` and compare those. If all of the keys are identical, then you can iterate through one set of them and do `g_hash_table_lookup(table1, key)` and `g_hash_table_lookup(table2, key)` and compare the values with `modulemd_hash_table_sets_are_equal()`.